### PR TITLE
Fix MI-536:Invalidate registry cache upon update on properties file

### PR DESCRIPTION
**Pull request by chanikag**
## Purpose
This is to fix https://github.com/wso2/micro-integrator/issues/536

MicroIntegratorRegistryListener is registered to monitor file changes in the registry.

### CODE REVIEW

File: components/mediation/registry/org.wso2.micro.integrator.registry/src/main/java/org/wso2/micro/integrator/registry/MicroIntegratorRegistryListener.java [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+        });
+    }
+
+    public void run() {
+        try {
+            WatchKey key;
+            while ((key = watcher.take()) != null) {
+                Path dir = keys.get(key);
+                if (dir == null) {
+                    log.error("WatchKey for listening to registry resources changes not recognized!!");
```
Comment (sdkottegoda) [inline components/mediation/registry/org.wso2.micro.integrator.registry/src/main/java/org/wso2/micro/integrator/registry/MicroIntegratorRegistryListener.java:92]: Do we need to do something like invalidating the key or something? Basically will the listener recover if it comes to this state?

File: components/mediation/registry/org.wso2.micro.integrator.registry/src/main/java/org/wso2/micro/integrator/registry/MicroIntegratorRegistry.java [hunk 2]
```diff
@@ -115,6 +121,34 @@ public MicroIntegratorRegistry() {
...
         initSecurityRepo();
+        initRegistryListener(defaultFSRegRoot);
+    }
+
+    private void initRegistryListener(String regRoot) {
+        try {
+            MicroIntegratorRegistryListener watcher = new MicroIntegratorRegistryListener(Paths.get(regRoot));
+            ExecutorService executor = Executors.newSingleThreadExecutor();
+            executor.submit(watcher);
+            executor.shutdown();
```
Comment (sdkottegoda) [inline components/mediation/registry/org.wso2.micro.integrator.registry/src/main/java/org/wso2/micro/integrator/registry/MicroIntegratorRegistry.java:132]: Is it correct to shutdown the executor here?

File: components/mediation/registry/org.wso2.micro.integrator.registry/src/main/java/org/wso2/micro/integrator/registry/MicroIntegratorRegistry.java [hunk 3]
```diff
@@ -115,6 +121,34 @@ public MicroIntegratorRegistry() {
...
+        } catch (IOException e) {
+            log.error(e);
```
Comment (sdkottegoda) [inline components/mediation/registry/org.wso2.micro.integrator.registry/src/main/java/org/wso2/micro/integrator/registry/MicroIntegratorRegistry.java:134]: Don't we need to throw an exception and interrupt the server startup?

File: components/mediation/registry/org.wso2.micro.integrator.registry/src/main/java/org/wso2/micro/integrator/registry/MicroIntegratorRegistryListener.java (hunk 1 above)
Comment (chanikag) [reply components/mediation/registry/org.wso2.micro.integrator.registry/src/main/java/org/wso2/micro/integrator/registry/MicroIntegratorRegistryListener.java:92]: Yes, the task is continued after that.. So the listener will take another key when an event happen.

File: components/mediation/registry/org.wso2.micro.integrator.registry/src/main/java/org/wso2/micro/integrator/registry/MicroIntegratorRegistry.java (hunk 2 above)
Comment (chanikag) [reply components/mediation/registry/org.wso2.micro.integrator.registry/src/main/java/org/wso2/micro/integrator/registry/MicroIntegratorRegistry.java:132]: Yes, this is to indicate that there is no task anymore to submit to the executor

File: components/mediation/registry/org.wso2.micro.integrator.registry/src/main/java/org/wso2/micro/integrator/registry/MicroIntegratorRegistry.java (hunk 3 above)
Comment (chanikag) [reply components/mediation/registry/org.wso2.micro.integrator.registry/src/main/java/org/wso2/micro/integrator/registry/MicroIntegratorRegistry.java]: +1. Fixed

File: components/mediation/registry/org.wso2.micro.integrator.registry/src/main/java/org/wso2/micro/integrator/registry/MicroIntegratorRegistry.java [hunk 4]
```diff
@@ -367,7 +403,7 @@ public RegistryEntry getRegistryEntry(String key) {
             entryEmbedded.setType(MicroIntegratorRegistryConstants.FILE);
             entryEmbedded.setDescription("Resource at : " + url.toString());
             entryEmbedded.setLastModified(urlc.getLastModified());
-            entryEmbedded.setVersion(urlc.getLastModified());
+            entryEmbedded.setVersion(getLastModifiedTimestamp(resolveRegistryURI(key), urlc));
```
Comment (Vathsan) [inline components/mediation/registry/org.wso2.micro.integrator.registry/src/main/java/org/wso2/micro/integrator/registry/MicroIntegratorRegistry.java:406]: Would only setting the version is enough? Do not we need to remove the particular key from [synapseConfiguration](https://github.com/wso2/wso2-synapse/blob/e01af546918c60e846fea1898fcc91431b19e90f/modules/core/src/main/java/org/apache/synapse/config/SynapseConfiguration.java#L929)?

Comment (chanikag) [reply components/mediation/registry/org.wso2.micro.integrator.registry/src/main/java/org/wso2/micro/integrator/registry/MicroIntegratorRegistry.java:406]: Yes, it checks for the version https://github.com/wso2/wso2-synapse/blob/master/modules/core/src/main/java/org/apache/synapse/registry/AbstractRegistry.java#L96

**Pull request merged**